### PR TITLE
Adding Engine

### PR DESCRIPTION
**Pull request by n3wscott**
Engine  is an optional component for client to leverage to start and stop the protocol impl.

- Adding context to End
- broke the benchmark for http, need help fixing it.

Signed-off-by: Scott Nichols <snichols@vmware.com>

### CODE REVIEW

File: cmd/samples/httpb/receiver-direct/main.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+func main() {
+	ctx := context.Background()
+	p, err := cloudevents.NewHTTP()
+	if err != nil {
+		log.Fatalf("failed to create protocol: %s", err.Error())
+	}
+
+	h, err := cloudevents.NewHTTPReceiveHandler(ctx, p, receive)
```
Comment (slinkydeveloper) [inline cmd/samples/httpb/receiver-direct/main.go]: That's coool!

File: pkg/transport/requester.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+	"github.com/cloudevents/sdk-go/pkg/binding"
+)
+
+// Requester sends a message and receives a response
+//
+// Optional interface that may be implemented by protocols that support
+// request/response correlation.
+type Requester interface {
+	// Request sends m like Sender.Send() but also arranges to receive a response.
```
Comment (slinkydeveloper) [inline pkg/transport/requester.go]: Why did you removed the constraint on implementing Sender?

File: test/benchmark/e2e/http/main.go [hunk 3]
```diff
@@ -156,28 +152,29 @@ func benchmarkReceiverSender(cases []e2e.BenchmarkCase, requestFactory func([]by
...
-			go func(client cloudevents.Client) {
-				_ = client.Send(ctx, e)
-				wg.Done()
-			}(clients[i])
-		}
-		wg.Wait()
-		return nil, nil
-	}}
-}
+//
```
Comment (slinkydeveloper) [inline test/benchmark/e2e/http/main.go]: I'll take care of it when this PR is merged

File: pkg/transport/requester.go (hunk 2 above)
Comment (n3wscott) [reply pkg/transport/requester.go]: You might want to only support request and not send

File: cmd/samples/httpb/receiver-direct/main.go (hunk 1 above)
Comment (n3wscott) [reply cmd/samples/httpb/receiver-direct/main.go]: I agree. I can see it working with the queues too

File: test/benchmark/e2e/http/main.go (hunk 3 above)
Comment (n3wscott) [reply test/benchmark/e2e/http/main.go]: Thanks. I tired but it is complex code

**Comment by n3wscott** [comment]
Will rename engine in a followup

**Pull request merged**